File: day 2/part2.py
from aocd import get_data
data = get_data(day=2, year=2024)

def check(line, dif):
    removed = False
    i = 1
    while i < len(line):
        if not (1 <= dif(line[i], line[i-1]) <= 3):
            # Either i-1 or i shouldn't be here
            if removed:
                return False

            # removing i-1 means we need to check i-2 works with i, and i works with i+1, then we add 2 to i
            # removing i means we need to check i-1 works with i+1, and then we add 2 to i
            # adding 2 gets done because of the i+1 outside the if

            # removing i-1
            if (i == 1 or 1 <= dif(line[i], line[i-2]) <= 3) and (i == len(line)-1 or 1 <= dif(line[i+1], line[i]) <= 3):
                i += 1
            # removing i
            elif (i == len(line)-1 or 1 <= dif(line[i+1], line[i-1]) <= 3):
                i += 1
            else:
                # can't remove either
                return False

            removed = True
        
        i += 1

    return True


total = 0
for line in data.split('\n'):
    line = [int(x) for x in line.split()]

    # Check with increasing and decreasing
    if check(line, lambda x, y: x - y) or check(line, lambda x, y: y - x):
        total += 1

print(total)




### BAD CODE ###
# def is_safe(line):
#     # Try increasing
#     for i in range(1, len(line)):
#         if line[i-1] >= line[i]:
#             break

#         if not (1 <= line[i] - line[i-1] <= 3):
#             break
#     else:
#         return True
    
#     # Try decreasing
#     for i in range(1, len(line)):
#         if line[i-1] <= line[i]:
#             break

#         if not (1 <= line[i-1] - line[i] <= 3):
#             break
#     else:
#         return True
    
#     return False

# total = 0
# for line in data.split('\n'):
#     line = [int(x) for x in line.split()]

#     for i in range(len(line)):
#         if is_safe(line[:i] + line[i+1:]):
#             total += 1
#             break

# def dif(n1, n2, first):
#     if first == 0:
#         return n1 - n2
#     return n2 - n1

File: day 2/part2.hs
import System.IO
import Data.Bool (Bool(True))

part2 :: String -> Int
part2 inp = sum [1 | line <- lines inp, is_safe line]
    where
        is_safe line = is_working ([(read x :: Int) | x <- words line]) (\x -> \y -> y - x) || is_working ([(read x :: Int) | x <- words line]) (\x -> \y -> x - y)

        is_working [] dif = True
        is_working [x] dif = True
        is_working [x, y] dif = True
        is_working (x:y:z:xs) dif
            | 1 <= dif x y && dif x y <= 3 = is_working' (x:y:z:xs) False dif -- first 2 items are fine, continue
            | 1 <= dif y z  && dif y z <= 3 = is_working' (y:z:xs) True dif -- removing first item
            | 1 <= dif x z && dif x z <= 3 = is_working' (x:z:xs) True dif -- removing second item
            | otherwise = False -- can't remove either item

        
        is_working' (x:y:[]) r dif
            | 1 <= dif x y && dif x y <= 3 = True
            | r = False
            | otherwise = True
        is_working' (w:x:y:[]) r dif
            | 1 <= dif x y && dif x y <= 3 = True
            | r = False
            | otherwise = True
        is_working' (w:x:y:z:xs) r dif -- we say y = whats at index i in python
            | 1 <= dif x y && dif x y <= 3 = is_working' (x:y:z:xs) r dif -- x and y are fine, continue
            | r = False -- we have to remove either x or y, but we have already removed something
            | (1 <= dif w y && dif w y <= 3) && (1 <= dif y z && dif y z <= 3) = is_working' (y:z:xs) True dif -- remove x
            | (1 <= dif x z && dif x z <= 3) = is_working' (x:z:xs) True dif -- remove y
            | otherwise = False -- can't remove either item


main :: IO ()
main = do
    inp <- readFile "input.txt"
    print (part2 inp)